Correctly record the dispose method to be invoked in operations

chsienki opened the pull request:
- Correctly record the dispose method in operations that will be used when not interface based
- Fix CFG to correctly handle struct based disposal and list the correct method
- Update and add tests


Fixes https://github.com/dotnet/roslyn/issues/49267

chsienki commented:
@dotnet/roslyn-compiler for review

Note: this is targeting @333fred's IOperation refactoring branch rather than master to reduce merge issues

333fred commented on src/Compilers/Core/Portable/Operations/OperationInterfaces.xml:
```diff
@@ -485,6 +485,11 @@
         </summary>
       </Comments>
     </Property>
+    <Property Name="DisposeMethod" Type="IMethodSymbol?">
```
It doesn't feel like making this only be populated if it's not `IDisposable.Dispose` is particularly useful, it feels more like an implementation detail bleeding through. I think we should just always populate this, and if it's `IDiposable` then we populate it with `IDisposable.Dispose`.

333fred commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3862,7 +3862,7 @@ void processResource(IOperation resource, ArrayBuilder<(IVariableDeclarationOper
                 LeaveRegion();
 
-                AddDisposingFinally(resource, knownToImplementIDisposable: true, iDisposable, isAsynchronous);
+                AddDisposingFinally(resource, knownToImplementIDisposable: true, iDisposable, disposeMethod, isAsynchronous);
```
Well, it's certainly not know to implement `IDisposable` if it's a ref struct. Please update the comments above and the parameter name.

333fred reviewed (commented):
Done review pass (commit 2)

chsienki commented:
@333fred  / @AlekseyTs assuming tests are good, this should be good to review now.

By always populating the dispose method in binding I was able to remove a bunch of logic in the operation and lowering to just use what we found, instead of calculating it again.

chsienki commented:
Note to self: run VB tests locally too. 🤦‍♂️

333fred commented:
Note on the determinism test: I haven't merged Aleksey's fix from master into this branch yet, so it's flaky.

333fred commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3813,23 +3813,40 @@ bool shouldConvertToIDisposableBeforeTry(IOperation resource)
+                // }
                 // Otherwise, when Resource type is a nullable value type or
-                // a reference type other than dynamic, the expansion is:
+                // a reference type other than dynamic, that implements IDisposable the expansion is:
```
>c, t [](start = 53, length = 4)

Nit: comma should go after the new addition

333fred commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3813,23 +3813,40 @@ bool shouldConvertToIDisposableBeforeTry(IOperation resource)
                 // }
                 //
+                // Otherwise, when Resource type is a nullable value type or
+                // a reference type other than dynamic, that implements disposal
```
>c, [](start = 53, length = 2)

Nit: comma should go after "via pattern"

333fred commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3888,61 +3905,65 @@ private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisp
-            AppendNewBlock(endOfFinally);
+                var isPatternDispose = !disposeMethod.ContainingType.Equals(iDisposable, SymbolEqualityComparer.Default);
+                var isNull = resource.GetConstantValue() == ConstantValue.Null;
+                var requiresConversion = !isPatternDispose && !isAsynchronous && !isNull && !_compilation.HasImplicitConversion(resource.Type, iDisposable) && !isNonNullableValueType(resource.Type);
```
>!_compilation.HasImplicitConversion(resource.Type, iDisposable) [](start = 92, length = 63)

If it has an implicit conversion, we probably still need to represent that implicit conversion, right? Or am I misunderstanding the expansion?

chsienki replied:
If it has an implicit conversion we'll represent that later on. This checks if we're doing a runtime conversion check. That only happens in foreach on enumerators. The simplest example is with a dynamic enumerator: we try and convert it to IDisposable first then check it for null. I'll give the var a more descriptive name.

333fred commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3888,61 +3905,65 @@ private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisp
+                if (requiresConversion)
+                {
+                    Debug.Assert(!isNonNullableValueType(resource.Type));
+                    resource = ConvertToIDisposable(resource, iDisposable, isTryCast: true);
```
This doesn't appear to match the expansion above: this says try-cast, but that is unconditional.

333fred commented:
Done review pass (commit 9)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationInterfaces.xml:
```diff
@@ -485,6 +485,11 @@
         </summary>
       </Comments>
     </Property>
+    <Property Name="DisposeMethod" Type="IMethodSymbol?">
```
><Property Name="DisposeMethod" Type="IMethodSymbol?"> [](start = 4, length = 53)

With this design choice we won't be able to represent different methods for different resources. This is probably sufficient today because only C# supports pattern based disposing and it cannot have resources of different types. However, VB can have resources of different types and, if we decide to support pattern based disposing in VB, we will have a hard time representing that in the API. Consider accounting for that in the shape of the API. For example, instead of a property, we could have a method that returns a MethodSymbol for specific resource identified either by index, or by specific node, representing the resource (IVariableDeclarationOperation or the singleton).

AlekseyTs replied:
Or the property can return an array of method symbols, but that would force us to always allocate an array which today will always be filled with the same value.   

---
In reply to: [528014394](https://github.com/dotnet/roslyn/pull/49481#discussion_r528014394) [](ancestors = 528014394)

chsienki replied:
That's interesting. I guess I figured we were safe from a VB point of view because we've said we wouldn't evolve it further. Agree about the array case. 

How would you envisage the method API working? That we'd have a method on the operation itself that returns the dispose method for each possible child resource?

AlekseyTs replied:
> How would you envisage the method API working? That we'd have a method on the operation itself that returns the dispose method for each possible child resource? 

Yes, something like that.

---
In reply to: [528022713](https://github.com/dotnet/roslyn/pull/49481#discussion_r528022713) [](ancestors = 528022713)

chsienki replied:
@AlekseyTs Hmm, I started playing around with this, and I think it's probably do-able, but requires a bit of re-layering. Given that we didn't previously expose these publically, would it be acceptable to just make the `DisposeMethod` `internal` for this PR, and file a an issue / follow up PR to expose it via a suitable public API?

AlekseyTs replied:
> Given that we didn't previously expose these publically, would it be acceptable to just make the DisposeMethod internal for this PR, and file a an issue / follow up PR to expose it via a suitable public API?

I think that would be the right thing to do at the moment. We are not exposing this information publicly for foreach, for example. 

---
In reply to: [528029327](https://github.com/dotnet/roslyn/pull/49481#discussion_r528029327) [](ancestors = 528029327)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationInterfaces.xml:
```diff
@@ -2934,6 +2939,11 @@
         <summary>True if this is an asynchronous using declaration.</summary>
       </Comments>
     </Property>
+    <Property Name="DisposeMethod" Type="IMethodSymbol?">
```
><Property Name="DisposeMethod" Type="IMethodSymbol?"> [](start = 4, length = 53)

Similar comment here

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3888,61 +3905,65 @@ private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisp
-            AppendNewBlock(endOfFinally);
+                var isPatternDispose = !disposeMethod.ContainingType.Equals(iDisposable, SymbolEqualityComparer.Default);
+                var isNull = resource.GetConstantValue() == ConstantValue.Null;
+                var requiresRuntimeConversion = !isPatternDispose && !isAsynchronous && !isNull && !_compilation.HasImplicitConversion(resource.Type, iDisposable) && !isNonNullableValueType(resource.Type);
```
>var requiresRuntimeConversion = !isPatternDispose && !isAsynchronous && !isNull && !_compilation.HasImplicitConversion(resource.Type, iDisposable) && !isNonNullableValueType(resource.Type); [](start = 16, length = 189)

This line is too long, consider splitting it.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3888,61 +3905,65 @@ private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisp
+                                          : _compilation.GetSpecialType(SpecialType.System_IDisposable);
 
-            AppendNewBlock(endOfFinally);
+                var isPatternDispose = !disposeMethod.ContainingType.Equals(iDisposable, SymbolEqualityComparer.Default);
```
>Default [](start = 112, length = 7)

I think it will be more robust long term to use comparer that ignores all possible aspects here.

AlekseyTs commented on src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -3,6 +3,8 @@ Microsoft.CodeAnalysis.GeneratorAttribute.Languages.get -> string[]
 Microsoft.CodeAnalysis.ITypeSymbol.IsRecord.get -> bool
 Microsoft.CodeAnalysis.Operations.OperationWalker<TArgument>
 Microsoft.CodeAnalysis.Operations.OperationWalker<TArgument>.OperationWalker() -> void
+Microsoft.CodeAnalysis.Operations.IUsingDeclarationOperation.DisposeMethod.get -> Microsoft.CodeAnalysis.IMethodSymbol
```
Should these lines be removed now?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3888,61 +3905,69 @@ private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisp
+                                          : _compilation.GetSpecialType(SpecialType.System_IDisposable);
 
-            AppendNewBlock(endOfFinally);
+                var isPatternDispose = !disposeMethod.ContainingType.Equals(iDisposable, SymbolEqualityComparer.ConsiderEverything);
```
>ConsiderEverything [](start = 112, length = 18)

This doesn't look like the comparer I suggested to use:

>  to use comparer that ignores all possible aspects here

chsienki replied:
It would help if I read what you actually wrote :)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -1179,8 +1179,7 @@ Namespace Microsoft.CodeAnalysis.Operations
                                                      statementInfo.MoveNextMethod,
                                                      isAsynchronous:=False,
                                                      boundForEachStatement.EnumeratorInfo.NeedToDispose,
-                                                     knownToImplementIDisposable:=boundForEachStatement.EnumeratorInfo.NeedToDispose AndAlso
```
>boundForEachStatement.EnumeratorInfo.NeedToDispose [](start = 82, length = 50)

I am not comfortable with us trying to infer this value later on. Code that calculates it is complicated enough and I would prefer us to not take any chances.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3888,61 +3905,69 @@ private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisp
+                var isNull = resource.GetConstantValue() == ConstantValue.Null;
+                var requiresRuntimeConversion = !isPatternDispose
+                                                && !isAsynchronous
+                                                && !isNull
```
>!isNull [](start = 51, length = 7)

It is not clear to me where this null check came from.

chsienki replied:
The VB version of `Using Nothing ... End Using` ends up putting the type of the operation as `Object`, meaning the logic here was attempting to perform a runtime check for it. 

As we're restoring the `knownToImplementIDisposable` members, it's not needed anyway. Agree we should be trying to calculate the minimal amount of things for this exact reason.

AlekseyTs replied:
> The VB version of Using Nothing ... End Using ends up putting the type of the operation as Object, meaning the logic here was attempting to perform a runtime check for it. 

Is addition of this check going to cause a behavior change for C# scenarios? This is exactly the kind of issue that I don't want us to deal here. 

---
In reply to: [528045939](https://github.com/dotnet/roslyn/pull/49481#discussion_r528045939) [](ancestors = 528045939)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3877,7 +3894,7 @@ void processResource(IOperation resource, ArrayBuilder<(IVariableDeclarationOper
             }
         }
 
-        private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisposable, ITypeSymbol iDisposable, bool isAsynchronous)
```
>knownToImplementIDisposable [](start = 67, length = 27)

I am not comfortable with elimination of this parameter. There is enough complexity in how it was calculated across both languages that I would prefer us to not take any chances.

chsienki replied:
I can restore it. I originally removed it because the presence of DisposeMethod should be enough to infer it, but I see no harm in being explicit.

chsienki replied:
The other issue was the fact that it doesn't really do what it says anymore. I think we should keep it but re-name it to `requiresRuntimeCheck` because that's all it was really being used for (in the getEnumerator case)

AlekseyTs replied:
>  I originally removed it because the presence of DisposeMethod should be enough to infer it.

Perhaps, that is the case. but how can we be sure that we re-inferred it with 100% accuracy?

---
In reply to: [528044035](https://github.com/dotnet/roslyn/pull/49481#discussion_r528044035) [](ancestors = 528044035)

AlekseyTs replied:
> The other issue was the fact that it doesn't really do what it says anymore. I think we should keep it but re-name it to requiresRuntimeCheck because that's all it was really being used for (in the getEnumerator case)

Renaming it would be fine, of course.

---
In reply to: [528044325](https://github.com/dotnet/roslyn/pull/49481#discussion_r528044325) [](ancestors = 528044325)

AlekseyTs replied:
I am not saying it is impossible to do. I am simply having a hard time proving to myself that we did it without causing a behavior change for some non-obvious scenario.

---
In reply to: [528044366](https://github.com/dotnet/roslyn/pull/49481#discussion_r528044366) [](ancestors = 528044366,528044035)

chsienki replied:
Absolutely. I restored the `knownToImplementIDisposable` and added hold of `IsPattern` too. Then in the builder we have a simple `RuntimeConversionRequired`

Much clearer what's going on and less opportunity to get it wrong.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/Loops/ForEachLoopOperationInfo.cs:
```diff
@@ -19,7 +19,7 @@ internal class ForEachLoopOperationInfo
 
         public readonly bool IsAsynchronous;
         public readonly bool NeedsDispose;
-        public readonly bool KnownToImplementIDisposable;
```
>public readonly bool KnownToImplementIDisposable; [](start = 8, length = 49)

I am not comfortable with elimination of this field. There is enough complexity in how it was calculated across both languages that I would prefer us to not take any chances with trying to re-infer it later.

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IUsingStatement.cs:
```diff
@@ -1962,43 +1955,22 @@ class NotDisposable
             Predecessors (0)
-            Statements (0)
-            Jump if True (Regular) to Block[B7]
-                IIsNullOperation (OperationKind.IsNull, Type: System.Boolean, IsInvalid, IsImplicit) (Syntax: 'GetDisposable() ?? input')
```
>IIsNullOperation (OperationKind.IsNull, Type: System.Boolean, IsInvalid, IsImplicit) (Syntax: 'GetDisposable() ?? input') [](start = 16, length = 121)

It appears that there are no changes like that for VB tests. Are we not testing similar error scenarios for VB?

chsienki replied:
There were no tests, I've added a couple to cover the scenarios and ensure they are as expected.

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IUsingStatement.cs:
```diff
@@ -1962,43 +1955,22 @@ class NotDisposable
         Block[B5] - Block
             Predecessors (0)
-            Statements (0)
-            Jump if True (Regular) to Block[B7]
```
>Jump if True (Regular) to Block[B7] [](start = 12, length = 35)

I would be more comfortable with the change if it didn't have any observable effect on scenarios that weren't actually broken before. Please consider restoring original error recovery logic.

chsienki replied:
I think I would argue that this *was* broken though. We were returning a CFG that said we would call `IDisposable.Dispose()` on something that we know isn't disposable.

AlekseyTs replied:
> I think I would argue that this was broken though. We were returning a CFG that said we would call IDisposable.Dispose() on something that we know isn't disposable.

This is an error recovery and the IInvocationOperation actually has all the right types in place, receiver is System.IDisposable. There is a conversion that would fail at runtime, but I don't see any harm in its presence because IConversionOperation doesn't imply that conversion succeeds. In fact, we even say that it is IsInvalid. I would argue that  the graph actually reflect user's intent as compiler "understood" it.

---
In reply to: [528057465](https://github.com/dotnet/roslyn/pull/49481#discussion_r528057465) [](ancestors = 528057465)

chsienki replied:
Makes sense. I've restored it back to how it was. The only changes now are for IAsync disposal which was clearly incorrect.

chsienki commented:
Thanks @AlekseyTs 

I think I've addressed your feedback: I've restored the recovery logic to how it was and added some tests to cover missing VB scenarios. Now the only changes we see to the tests are when using `IAsyncDisposable` and and updated error recovery for the corner case when `IDisposable` is missing completely.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3877,7 +3894,7 @@ void processResource(IOperation resource, ArrayBuilder<(IVariableDeclarationOper
         }
 
-        private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisposable, ITypeSymbol iDisposable, bool isAsynchronous)
+        private void AddDisposingFinally(IOperation resource, IMethodSymbol? disposeMethod, bool isAsynchronous, bool requiresRuntimeConversion = false)
```
>requiresRuntimeConversion [](start = 118, length = 25)

Please make it a required parameter. Its value is significant and everybody (a dev and reviewers) should pay special attention to it.

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IUsingStatement.cs:
```diff
@@ -2964,37 +2926,23 @@ void M(System.IDisposable input, object o)
     {
         Block[B4] - Block
             Predecessors (0)
-            Statements (0)
```
>Statements (0) [](start = 12, length = 14)

A change for this scenario is unexpected

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IForEachLoopStatement.cs:
```diff
@@ -4844,21 +4844,13 @@ static class Extensions
-                        (ImplicitReference)
-                      Operand: 
-                        IFlowCaptureReferenceOperation: 0 (OperationKind.FlowCaptureReference, Type: System.Collections.Generic.IAsyncEnumerator<System.String>, IsImplicit) (Syntax: 'new Program()')
+            Statements (0)
```
>Statements (0) [](start = 12, length = 14)

Does the change here match the IL that we would emit?

chsienki replied:
This is async, so the IL we emit looks nothing like this. This is however the correct representation of what we 'would' emit before lowering to the state machine happens.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3888,61 +3905,70 @@ private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisp
+                              )?.GetISymbol() as IMethodSymbol;
 
-            if (!iDisposable.Equals(resource.Type))
+            if (disposeMethod is object)
```
>if (disposeMethod is object) [](start = 12, length = 28)

It is quite possible that I am missing something obvious, but it feels to me that the code in this method was rewritten without actual necessity to do that. For example, addition of this ```if``` looks unnecessary, we handled the situation when there is no method symbol before this change.  It feels to me that all what was needed is just a small change to the tryDispose local function and to the ```if``` above its invocation, which is responsible for the conversion of the resource. I suggest to try starting from the original state of the function and making minimal changes only if necessary.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -843,8 +843,10 @@ private EnumeratorResult GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder bui
+                // We don't know the runtime type, so we will have to insert a runtime check for disposal
+                // that will call the interface method if the check is successful
                 builder.NeedsDisposal = true;
+                builder.DisposeMethod = this.Compilation.GetWellKnownDisposeMethod(isAsync);
```
>this.Compilation.GetWellKnownDisposeMethod(isAsync); [](start = 40, length = 52)

We are locating a method without reporting any use site diagnostics. Why do we need to do this? I think there is no real need for us to pass method symbols for non-pattern-base disposal. Flow graph builder knows how to find them. So I wouldn't expect us to make any changes related to non-pattern-based disposal

chsienki replied:
The idea behind this change was the CFG wouldn't do the finding of the method. Instead we make all the logic decisions in the binder once, and CFG/IOperation just reports what was found during binding. If we keep finding the method later on we end up smearing out the logic in multiple places, and end up with bugs like the one this PR is aiming to fix, where the logic isn't updated in all the various places it's used.

AlekseyTs replied:
> and end up with bugs like the one this PR is aiming to fix, where the logic isn't updated in all the various places it's used.

I do not believe this is the case. It is enough for us to record the method that was found based on a pattern. We can deal interface based disposal the same way we did before. A null value is good enough to represent the scenario and all the interested parties know how to deal with interface-based disposal. I do not see any good reason to make any changes around these scenarios. The risk to break things is high and the benefit is very small in my opinion.  

---
In reply to: [533647453](https://github.com/dotnet/roslyn/pull/49481#discussion_r533647453) [](ancestors = 533647453)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -924,6 +926,7 @@ private void GetDisposalInfoForEnumerator(ref ForEachEnumeratorInfo.Builder buil
                     ref useSiteDiagnostics).IsImplicit)
             {
                 builder.NeedsDisposal = true;
+                builder.DisposeMethod = this.Compilation.GetWellKnownDisposeMethod(isAsync);
```
> this.Compilation.GetWellKnownDisposeMethod(isAsync); [](start = 39, length = 53)

Same comment.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -992,8 +996,10 @@ private ForEachEnumeratorInfo.Builder GetDefaultEnumeratorInfo(ForEachEnumerator
+            // We don't know the runtime type, so we will have to insert a runtime check for disposal
+            // that will call the interface method if the check is successful
             builder.NeedsDisposal = true;
+            builder.DisposeMethod = this.Compilation.GetWellKnownDisposeMethod(isAsync);
```
>this.Compilation.GetWellKnownDisposeMethod(isAsync); [](start = 36, length = 52)

Same comment

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -1181,6 +1181,8 @@ Namespace Microsoft.CodeAnalysis.Operations
                                                      boundForEachStatement.EnumeratorInfo.NeedToDispose,
                                                      knownToImplementIDisposable:=boundForEachStatement.EnumeratorInfo.NeedToDispose AndAlso
                                                                                   boundForEachStatement.EnumeratorInfo.IsOrInheritsFromOrImplementsIDisposable,
+                                                     disposeMethod:=DirectCast(DirectCast(_semanticModel.Compilation, VisualBasicCompilation).GetSpecialTypeMember(SpecialMember.System_IDisposable__Dispose), MethodSymbol),
```
>DirectCast(DirectCast(_semanticModel.Compilation, VisualBasicCompilation).GetSpecialTypeMember(SpecialMember.System_IDisposable__Dispose), MethodSymbol) [](start = 68, length = 152)

I think we should simple pass Nothing here and let flow graph builder figure the symbol out

chsienki replied:
But then we end up in the same situation where CFG has to re-infer decisions that the binding phase already knew. We'll probably never change it for VB so it makes sense that VB says 'this is the method I will use'. C# is then free to extend the disposal rules further without us having to special case it in the CFG builder.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -1435,7 +1437,8 @@ Namespace Microsoft.CodeAnalysis.Operations
             Dim isImplicit As Boolean = boundUsingStatement.WasCompilerGenerated
-            Return New UsingOperation(resources, body, locals, isAsynchronous:=False, _semanticModel, syntax, isImplicit)
+            Dim disposeMethod As MethodSymbol = DirectCast(DirectCast(_semanticModel.Compilation, VisualBasicCompilation).GetSpecialTypeMember(SpecialMember.System_IDisposable__Dispose), MethodSymbol)
+            Return New UsingOperation(resources, body, locals, isAsynchronous:=False, disposeMethod, _semanticModel, syntax, isImplicit)
```
>disposeMethod [](start = 86, length = 13)

Same comment

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs:
```diff
@@ -125,7 +125,7 @@ internal static BoundStatement BindUsingStatementOrDeclarationFromParts(SyntaxNo
                 {
                     iDisposableConversion = Conversion.ImplicitDynamic;
-                    disposeMethodOpt = null;
+                    disposeMethodOpt = originalBinder.Compilation.GetWellKnownDisposeMethod(isAsync: hasAwait);
```
>originalBinder.Compilation.GetWellKnownDisposeMethod(isAsync: hasAwait); [](start = 39, length = 72)

Same comment as for other call-sites

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs:
```diff
@@ -185,6 +185,7 @@ bool populateDisposableConversionOrDisposeMethod(bool fromExpression, out Conver
 
                 if (iDisposableConversion.IsImplicit)
                 {
+                    disposeMethodOpt = originalBinder.Compilation.GetWellKnownDisposeMethod(hasAwait);
```
>disposeMethodOpt = originalBinder.Compilation.GetWellKnownDisposeMethod(hasAwait); [](start = 20, length = 82)

The value is not used

AlekseyTs replied:
>originalBinder.Compilation.GetWellKnownDisposeMethod(hasAwait); [](start = 39, length = 63)

Same comment

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1646,18 +1646,24 @@ private IForLoopOperation CreateBoundForStatementOperation(BoundForStatement bou
+                                    : compilation.GetSpecialType(SpecialType.System_IDisposable);
+
+                var enumeratorImplementsDisposable = enumeratorInfoOpt.NeedsDisposal
+                                                     && compilation.Conversions.ClassifyImplicitConversionFromType(enumeratorInfoOpt.GetEnumeratorMethod.ReturnType,
```
>enumeratorInfoOpt.GetEnumeratorMethod [](start = 115, length = 37)

It looks like we dropped a null check for this value.

chsienki replied:
It was never supposed to be able to contain a null value. The enumeratorInfo has nullable enabled.

AlekseyTs replied:
> It was never supposed to be able to contain a null value. The enumeratorInfo has nullable enabled.

It looks to me that definition of ForEachEnumeratorInfo is in nullable disabled region.

---
In reply to: [533648164](https://github.com/dotnet/roslyn/pull/49481#discussion_r533648164) [](ancestors = 533648164)

AlekseyTs commented on src/Compilers/VisualBasic/Test/IOperation/IOperation/IOperationTests_IUsingStatement.vb:
```diff
@@ -1008,7 +1008,7 @@ IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null) (S
         <CompilerTrait(CompilerFeature.IOperation, CompilerFeature.Dataflow)>
         <Fact()>
-        Public Sub UsingFlow_14()
+        Public Sub UsingFlow_01()
```
>UsingFlow_01 [](start = 19, length = 12)

Why was that renamed? Is it possible that it corresponded to C# test with the same name?

chsienki replied:
Long ago there were `UsingFlow_{1-13}` in C# and then {`14,15}` in VB, but C# has since added significantly more tests with `{14...}` so it makes no sense to have a random set of numbered VB tests. We can move these to the end of the C# tests, but the likelihood is just that they'll clash again.

chsienki replied:
Ideally we'd give them descriptive names tbh

AlekseyTs replied:
> We can move these to the end of the C# tests, but the likelihood is just that they'll clash again. 

There is no requirement for test methods to be sorted in any way. The test was not affected by the change as well. I find the rename unnecessary.

---
In reply to: [533652564](https://github.com/dotnet/roslyn/pull/49481#discussion_r533652564) [](ancestors = 533652564)

AlekseyTs commented on src/Compilers/VisualBasic/Test/IOperation/IOperation/IOperationTests_IUsingStatement.vb:
```diff
@@ -1152,7 +1152,7 @@ Block[B8] - Exit
         <CompilerTrait(CompilerFeature.IOperation, CompilerFeature.Dataflow)>
         <Fact()>
-        Public Sub UsingFlow_15()
+        Public Sub UsingFlow_02()
```
>UsingFlow_02 [](start = 19, length = 12)

Why was this renamed?

AlekseyTs commented on src/Compilers/VisualBasic/Test/IOperation/IOperation/IOperationTests_IUsingStatement.vb:
```diff
@@ -1236,5 +1236,171 @@ Block[B6] - Exit
+
+        <CompilerTrait(CompilerFeature.IOperation, CompilerFeature.Dataflow)>
+        <Fact()>
+        Public Sub UsingFlow_04()
```
>UsingFlow_04 [](start = 19, length = 12)

Is there a similar C# test? Consider using the same name.

AlekseyTs commented on src/Compilers/VisualBasic/Test/IOperation/IOperation/IOperationTests_IUsingStatement.vb:
```diff
@@ -1236,5 +1236,171 @@ Block[B6] - Exit
 
+        <CompilerTrait(CompilerFeature.IOperation, CompilerFeature.Dataflow)>
+        <Fact()>
+        Public Sub UsingFlow_03()
```
>UsingFlow_03 [](start = 19, length = 12)

Is there a similar C# test? Consider using the same name.

AlekseyTs commented:
Done with review pass (commit 20). Overall, it feels to me that a lot of changes could be avoided. Let's discuss offline in more details.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3888,61 +3905,70 @@ private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisp
-                        return new AwaitOperation(invocation, semanticModel: null, value.Syntax, _compilation.GetSpecialType(SpecialType.System_Void), isImplicit: true);
-                    }
+                // resource.Dispose();
+                resource = new InvocationOperation(disposeMethod, resource, isVirtual: true,
```
>new InvocationOperation(disposeMethod, resource, isVirtual: true, [](start = 27, length = 65)

Can pattern-based expose use an extension method? In that scenario, the resource should be the first parameter instead. Do we need to handle omitted optional and param array parameters?

chsienki replied:
No, pattern disposal does not support extensions methods. We should report the params and optional arguments, however

AlekseyTs replied:
> We should report the params and optional arguments, however

See how we deal with arguments for GetEnumerator and MoveNext in VisitForEachLoop

---
In reply to: [533774968](https://github.com/dotnet/roslyn/pull/49481#discussion_r533774968) [](ancestors = 533774968)

chsienki replied:
@AlekseyTs I have a fix for this, but its a little involved, as we now need to pass the binder down through the `BoundUsingStatement` like we do for Foreach. Given the mild complexity, do you want me to add it to this PR, or wait and open a separate one?

AlekseyTs replied:
> Given the mild complexity, do you want me to add it to this PR, or wait and open a separate one?

I think a separate PR is fine.

---
In reply to: [534633638](https://github.com/dotnet/roslyn/pull/49481#discussion_r534633638) [](ancestors = 534633638)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3813,23 +3813,40 @@ bool shouldConvertToIDisposableBeforeTry(IOperation resource)
+                // {
+                //   ResourceType resource = expr;
+                //   try { statement; }
+                //   finally { if (resource != null) resource.Dispose(); }
```
>resource [](start = 53, length = 8)

Is this accurate for a nullable value type? We are not unwrapping the ```Nullable<T>```?

chsienki replied:
Updated the comment to show what we do in lowering.

AlekseyTs replied:
> Updated the comment to show what we do in lowering.

Why would you do this? Is this comment for lowering? Is this comment for lowering of a specific compiler?

---
In reply to: [534592569](https://github.com/dotnet/roslyn/pull/49481#discussion_r534592569) [](ancestors = 534592569)

chsienki replied:
I understood your comment to mean that we should be showing the unwrapping of the nullable in CFG, but clearly I misunderstood you.

AlekseyTs replied:
> I understood your comment to mean that we should be showing the unwrapping of the nullable in CFG, but clearly I misunderstood you.

I asked you about behavior that you are describing in a new comment specifically for a pattern-based case. I couldn't find a test for it, hence the question. Clearly the comment looked strange to me, but you started looking at completely different scenario.

---
In reply to: [534620372](https://github.com/dotnet/roslyn/pull/49481#discussion_r534620372) [](ancestors = 534620372)

chsienki replied:
Apologies, as I said, I misunderstood what you were asking.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:
```diff
@@ -963,7 +964,7 @@ private MethodSymbol GetGetEnumeratorMethod(NamedTypeSymbol collectionType, Diag
         }
 
-        private ForEachEnumeratorInfo.Builder GetDefaultEnumeratorInfo(ForEachEnumeratorInfo.Builder builder, DiagnosticBag diagnostics, TypeSymbol collectionExprType)
+        private ForEachEnumeratorInfo.Builder GetDefaultEnumeratorInfo(ForEachEnumeratorInfo.Builder builder, DiagnosticBag diagnostics, TypeSymbol collectionExprType, bool isAsync)
```
>, bool isAsync [](start = 166, length = 14)

It looks like this parameter is not used.

chsienki replied:
Removed.

AlekseyTs commented:
Done with review pass (commit 24)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3813,23 +3813,40 @@ bool shouldConvertToIDisposableBeforeTry(IOperation resource)
                 // }
                 //
+                // Otherwise, when ResourceType is a non-nullable value type that implements
+                // disposal via pattern, the expansion is:
```
Are we testing this scenario?

chsienki replied:
Yes in `UsingDeclaration_Flow_20`

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3813,23 +3813,40 @@ bool shouldConvertToIDisposableBeforeTry(IOperation resource)
                 //   finally { if (resource != null) ((IDisposable)resource).Dispose(); }
                 // }
                 //
+                // Otherwise, when Resource type is a nullable value type or
```
>when Resource type is a nullable value type [](start = 30, length = 43)

Are we testing this scenario?

chsienki replied:
Yes, in `UsingFlow_05` but actually, it doesn't appear to be correctly reporting what we do. It matches the comments in the Builder, but that _isnt_ how we lower the code, as you pointed out we unwrap the nullable, which we're not representing here. https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABATAAgAUBYAKAG8z9r9sAWfAWQApGBPAEQEsoAHAeygRgAGzgB+fFwB2vAK4wk+YP34jlASjIB6AFQAeYDJQA+Xdsqka12gAZ8zAOJwY3PoOFjmG/OMkz5GC0rG2pLUNDgfABefBgEOTgAbioIgF9U6jS9fW0jaVNzMkymTh4BIVEJfGdXco8q7xiTOIALBH4Ad3xpORERFNIM0jJYBIBjGFK3Cs84fBBaAEYAOgBJGYaxMnCabABmWgZNqDgm6JaYdq6evoGyNKA=

chsienki replied:
Oh, interesting. Actually we're representing in ControlFlowGraph what the spec says we *should* be doing: 
http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Lowering/LocalRewriter/LocalRewriter_UsingStatement.cs,277

Except in lowering we purposefully deviate from the spec to improve perf. 

@AlekseyTs what do you think? Should the CFG represent the spirit of the code in the spec, or what we *actually* lower to? I guess I lean towards representing the spec? because like the async stuff we give a general representation of what it's lowered to, not the implementation specific details.

AlekseyTs replied:
It looks to me that UsingFlow_05 and the sharplab link is about non-pattern disposal and this change is not supposed to make any changes for these scenarios. I am fine with the current behavior of the builder for them. The question I have is about pattern-based disposal for nullable value types. This is what this comment is about. Are we testing the scenario? 

---
In reply to: [534599322](https://github.com/dotnet/roslyn/pull/49481#discussion_r534599322) [](ancestors = 534599322)

chsienki replied:
It is not currently possible to perform pattern disposal on a nullable value types: see https://github.com/dotnet/roslyn/issues/34701 

I will add a test to show the error recovery scenario, and update the comment in CFG to correctly reflect that.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3813,23 +3813,48 @@ bool shouldConvertToIDisposableBeforeTry(IOperation resource)
+                // {
+                //   ResourceType resource = expr;
+                //   try { statement; }
+                //   finally { if (resource.HasValue()) ((IDisposable)resource.GetValueOrDefault()).Dispose(); }
```
>finally { if (resource.HasValue()) ((IDisposable)resource.GetValueOrDefault()).Dispose(); } [](start = 21, length = 91)

Where do you see this in flow graph tests?

chsienki replied:
Mentioned above. We don't the test shows something different: do we want to make the CFG represent the actual lowered value or what the spec says?

AlekseyTs replied:
> Mentioned above, we don't. The test shows something different: do we want to make the CFG represent the actual lowered value or what the spec says?

Does the comment accurately reflect CFG builder's behavior? If so, we don't want to make any changes for the comment. We also do not want to make any changes for scenarios that are not about pattern-based disposal, which is supposed to be the focus of this PR.

---
In reply to: [534616619](https://github.com/dotnet/roslyn/pull/49481#discussion_r534616619) [](ancestors = 534616619)

chsienki replied:
Yes apologies. I was confused by your comment, I thought you were saying that the comment, and thus the tests are incorrect for those scenarios.

AlekseyTs reviewed (approved):
LGTM (commit 26), assuming that we will deal with the arguments in a separate PR.

chsienki commented:
@dotnet/roslyn-compiler / @333fred for a second review please :)

333fred reviewed (approved):
LGTM (commit 26)
